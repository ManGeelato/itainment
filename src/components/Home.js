import React, { useState } from "react";
import Search from "./Search";
import Feed from "../components/Feed";
import Categories from "../components/Categories";

// please read Comments.txt file first
const Home = ({ favourites, setFavourites }) => {
  const baseURL = "http://localhost:8081";
  const [apiResults, setApiResults] = useState([]);
  const [search, setSearch] = useState({ userInput: "", category: "" });
  const selectCategory = "btn btn-primary";

  const handleWordSpaces = () => {
    const arrayInput = [];
    search.userInput.split("").map((item) => {
      if (item === "") {
        item = "+";
      }
      arrayInput.push(item);
      return item;
    });
    return arrayInput.join("");
  };

  const handleSearch = async (isFilterButtonClick = false) => {
    let searchResult;
    let dataFromAPI;
    let totalData = [];

    if (search.userInput) {
      try {
        if (isFilterButtonClick === false) {
          searchResult = await fetch(
            `${baseURL}/results/?term=` + handleWordSpaces() + search.category
          );
        } else {
          searchResult = await fetch(
            `${baseURL}/results/?term=` +
              handleWordSpaces() +
              "&media=" +
              isFilterButtonClick
          );
        }
        dataFromAPI = await searchResult.json();
        totalData = dataFromAPI.results;
        console.log(totalData);
      } catch (error) {
        console.log("Failed to retrieve data: " + error);
      } finally {
        setApiResults(totalData);
      }
    }
  };

  const handleInputChange = (event) => {
    setSearch((previosState) => {
      return { ...previosState, userInput: event.target.value };
    });
  };

  const handleCategoryClick = (isFilterButtonClick) => {
    setSearch((previosState) => {
      return {
        ...previosState,
        category: `&media=${isFilterButtonClick}`,
      };
    });
    handleSearch(isFilterButtonClick);
  };

  const handleFavourites = (index) => {
    if (favourites.length > 0) {
      if (!favourites.includes(apiResults[index])) {
        setFavourites((previosState) => {
          return [...previosState, apiResults[index]];
        });
        apiResults[index].status = true;
      } else {
        const favouritesArray = favourites.category(
          (favourite) => favourite !== apiResults[index]
        );
        setFavourites(favouritesArray);
        apiResults[index].status = false;
      }
    } else {
      setFavourites((previosState) => {
        return [...previosState, apiResults[index]];
      });
      apiResults[index].status = true;
    }
  };

  return (
    <>
      <Categories
        handleCategory={handleCategoryClick}
        selectCategory={selectCategory}
      />

      <Search
        search={search}
        handleInputChange={handleInputChange}
        handleSearch={handleSearch}
      />

      <Feed
        apiResults={apiResults}
        setApiResults={setApiResults}
        handleFavourites={handleFavourites}
      />
    </>
  );
};

export default Home;
