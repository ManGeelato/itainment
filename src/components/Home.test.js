import React from 'react';
import renderer from 'react-test-renderer';

import Home from './Home';

test('Is the component working as expected', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
})

describe("getMedia", () => {
    test("Is Media being fetched from API", async () => {
        const tree = renderer.create(<Home />).toJSON();
        expect(tree).toMatchSnapshot();
    })
})