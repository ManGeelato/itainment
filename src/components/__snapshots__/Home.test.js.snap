// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Is the component working as expected 1`] = `
Array [
  <main
    id="categories"
  >
    <h4>
      Categories
    </h4>
    <div
      className="form-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="movie"
        onClick={[Function]}
      >
        Movie
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="podcast"
        onClick={[Function]}
      >
        Podcast
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="music"
        onClick={[Function]}
      >
        Music
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="audio-book"
        onClick={[Function]}
      >
        Audio-book
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="short-film"
        onClick={[Function]}
      >
        Short-film
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="tv-show"
        onClick={[Function]}
      >
        TV-show
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="software"
        onClick={[Function]}
      >
        Software
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="e-book"
        onClick={[Function]}
      >
        E-book
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="all"
        onClick={[Function]}
      >
        All
      </label>
    </div>
  </main>,
  <form
    className="form-inline my-2 my-lg-0"
    id="search"
  >
    <input
      aria-label="Search"
      className="form-control"
      onChange={[Function]}
      placeholder="Search music, films, ebooks, audiobook, movies and many more..."
      type="search"
      value=""
    />
    <button
      className="btn btn-outline-success my-2 my-sm-0"
      id="btnSearch"
      onClick={[Function]}
      type="submit"
    >
      Search
    </button>
  </form>,
  <div
    className="card"
    id="feed"
  >
    Your search category will appear here...
  </div>,
]
`;

exports[`getMedia Is Media being fetched from API 1`] = `
Array [
  <main
    id="categories"
  >
    <h4>
      Categories
    </h4>
    <div
      className="form-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="movie"
        onClick={[Function]}
      >
        Movie
      </label>
    </div>
    <div
      className="form-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="podcast"
        onClick={[Function]}
      >
        Podcast
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="music"
        onClick={[Function]}
      >
        Music
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="audio-book"
        onClick={[Function]}
      >
        Audio-book
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="short-film"
        onClick={[Function]}
      >
        Short-film
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="tv-show"
        onClick={[Function]}
      >
        TV-show
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="software"
        onClick={[Function]}
      >
        Software
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="e-book"
        onClick={[Function]}
      >
        E-book
      </label>
    </div>
    <div
      className="htmlForm-check form-check-inline"
    >
      <label
        className="btn btn-primary"
        htmlFor="all"
        onClick={[Function]}
      >
        All
      </label>
    </div>
  </main>,
  <form
    className="form-inline my-2 my-lg-0"
    id="search"
  >
    <input
      aria-label="Search"
      className="form-control"
      onChange={[Function]}
      placeholder="Search music, films, ebooks, audiobook, movies and many more..."
      type="search"
      value=""
    />
    <button
      className="btn btn-outline-success my-2 my-sm-0"
      id="btnSearch"
      onClick={[Function]}
      type="submit"
    >
      Search
    </button>
  </form>,
  <div
    className="card"
    id="feed"
  >
    Your search category will appear here...
  </div>,
]
`;
